src/submission/submission-api.ts

```diff
@@ -88,9 +92,37 @@ class SubmissionController {
     await submission.operations.deleteRegistration(registrationId, programId);
     return res.status(200).send();
   }
+
+  @HasSubmittionAccess((req: Request) => req.params.programId)
+  async saveClinicalTsvFile(req: Request, res: Response) {
+    if (!isValidCreateBody(req, res, req.params.clinicalType)) {
+      return;
+    }
+    const file = req.file;
+    let records: ReadonlyArray<Readonly<{ [key: string]: string }>>;
+    try {
+      records = await TsvUtils.tsvToJson(file.path);
+    } catch (err) {
+      return ControllerUtils.badRequest(res, {
+        msg: `failed to parse the tsv file: ${err}`,
+        code: ErrorCodes.TSV_PARSING_FAILED
+      });
+    }
+    res.set("Content-Type", "application/json");
+    const command: SaveClinicalCommand = {
+      records: records,
+      programId: req.params.programId,
+      clinicalType: req.params.clinicalType
+    };
+    const result = await submission.operations.uploadClinical(command);
+    if (!result.successful) {
+      return res.status(422).send(result);
+    }
+    return res.status(200).send(result);
+  }
 }
 
-const isValidCreateBody = (req: Request, res: Response): boolean => {
+const isValidCreateBody = (req: Request, res: Response, type: FileType): boolean => {
```

Comment:
can you check if the type is undefined and check that it's valid type

Reply:
I see you're doing that but it's being used before in the method

Reply:
I moved the check before type is used. The check works even if type is undefined, it will return badRequest.